algorithms/integration/processor.py

```diff
@@ -792,43 +817,111 @@ def compute_processors(self):
+            available_swap = psutil.swap_memory().free
+            available_incl_swap = available_memory + available_swap
+            available_immediate_limit = (
+                available_memory * self.params.block.max_memory_usage
+            )
+            logger.debug(
+                "Available memory: %.1f GB, swap: %.1f GB, " "memory limit: %.1f GB",
+                available_memory / 1e9,
+                available_swap / 1e9,
+                available_immediate_limit / 1e9,
+            )
+            rlimit = getattr(resource, "RLIMIT_VMEM", getattr(resource, "RLIMIT_AS"))
+            if rlimit:
+                # try to check for process limit values
+                try:
+                    ulimit = resource.getrlimit(rlimit)[0]
+                    if ulimit <= 0:
+                        logger.debug("No memory ulimit set")
+                    else:
+                        ulimit_used = psutil.Process().memory_info().rss
+                        logger.debug(
+                            "Detected ulimit of %d bytes, of which %d are used",
+                            ulimit,
+                            ulimit_used,
+                        )
+                        available_memory = max(
+                            0, min(available_memory, ulimit - ulimit_used)
+                        )
+                        available_incl_swap = max(
+                            0, min(available_incl_swap, ulimit - ulimit_used)
+                        )
+                        available_immediate_limit = (
+                            available_memory * self.params.block.max_memory_usage
+                        )
+                        logger.debug(
+                            "Revised available memory: %.1f GB, "
+                            "revised available memory incl. swap: %.1f GB, "
+                            "revised immediate limit: %.1f GB",
+                            available_memory / 1e9,
+                            available_incl_swap / 1e9,
+                            available_immediate_limit / 1e9,
+                        )
+                except Exception as e:
+                    logger.debug("Could not obtain ulimit values due to %s", str(e))
+            njobs = available_immediate_limit / memory_required_per_process
+            if njobs >= self.params.mp.nproc:
+                # There is enough memory. Take no action
+                self.params.block.max_memory_usage /= self.params.mp.nproc  # why?
+                return
+            if njobs >= 1:
+                # There is enough memory to run, but not as many processes as requested
+                logger.info(
+                    "Reducing number of processes from %d to %d due to memory constraints",
+                    self.params.mp.nproc,
+                    int(njobs),
+                )
+                self.params.mp.nproc = int(njobs)
+                self.params.block.max_memory_usage /= self.params.mp.nproc  # why?
```

Comment:
Since you have made the effort to look at all this - `# why?` we should probably resolve as a part of this review?

Reply:
The second check (see above 'why duplicate?') happens at the start of each task. But each task is now only allowed `1/n` of the total memory, which is what I assume this is for.
The only way this makes any sense is when the task is run on a different machine, but in that case it doesn't make any sense :thinking:

In my opinion we should scratch the per-task check. Check once at the beginning of the process and either there is enough memory or there is not. We don't catch memory disappearing while we're running inside a task, so it's not like this additional check is of much use.

Reply:
+1 to scratching per-task check

Reply:
Done.